### PR TITLE
Make CrateMetadata and CStore thread-safe

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

File: src/librustc_metadata/creader.rs [hunk 1]
```diff
@@ -335,7 +334,9 @@ impl<'a> CrateLoader<'a> {
                 if data.root.macro_derive_registrar.is_some() {
                     dep_kind = DepKind::UnexportedMacrosOnly;
                 }
-                data.dep_kind.set(cmp::max(data.dep_kind.get(), dep_kind));
+                data.dep_kind.with_lock(|data_dep_kind| {
```
[the remark this answers is not in the document]
Comment (michaelwoerister) [reply src/librustc_metadata/creader.rs]: `parking_lot::Once` looks like a good building block for such a thing.

**Review by michaelwoerister — COMMENTED** [review]
Thanks, @Zoxc! Looks good to me. 

What do you think of adding an `InitOnce<T>` primitive to `rustc_data_structures::sync`? Especially things like `codemap_import_info` would profit from this, I think. I haven't done any performance testing but even if it's not significantly faster, it makes the semantics much clearer which is always a win.

**Comment by Zoxc** [comment]
I did think about adding something like `InitOnce`, but I think we'll end up using either queries (which cannot deadlock) or just refactoring away the internal mutability.

File: src/librustc_metadata/cstore.rs [hunk 2]
```diff
@@ -90,21 +89,23 @@ pub struct CrateMetadata {
...
-            metas: RefCell::new(IndexVec::new()),
-            extern_mod_crate_map: RefCell::new(FxHashMap()),
+            metas: RwLock::new(IndexVec::new()),
+            extern_mod_crate_map: Lock::new(FxHashMap()),
             metadata_loader,
         }
     }
 
+    /// You cannot use this function to allocate a CrateNum in a thread-safe manner.
+    /// It is currently only used in CrateLoader which is single-threaded code.
```
Comment (Zoxc) [inline src/librustc_metadata/cstore.rs]: I added a comment about this function. `next_crate_num` is used once in `CrateLoader` to get the next free `CrateNum`. `CrateLoader` may then create multiple `CrateNum`s based on that.

Comment (michaelwoerister) [reply src/librustc_metadata/cstore.rs]: :+1:

**Comment by michaelwoerister** [comment]
> I did think about adding something like InitOnce, but I think we'll end up using either queries (which cannot deadlock) or just refactoring away the internal mutability.

At least for `codemap_import_info` we won't want to use a query, I think.

**Comment by michaelwoerister** [comment]
@bors r+

Let's merge this as is, if you don't mind, @Zoxc. I'll be looking into the whole crate store infrastructure a bit more in detail over the next few days.

**Pull request merged**